contents/docs/how-posthog-works/ingestion-pipeline.mdx

```diff
@@ -0,0 +1,172 @@
+---
+title: Ingestion pipeline
+---
+
+In its simplest form, the PostHog ingestion pipeline is a collection of services which listen for events as they are sent in, process them, and then store them for later analysis.
+This document gives an overview of how data ingestion works, as well as some of the caveats to be aware of when sending events to PostHog.
+
+## Capture API
+
+The Capture API represents the user-facing side of the ingestion pipeline, and is exposed as a number of API routes where events can be sent.
+Before an event reaches the Ingestion pipeline, there are a couple of preliminary checks and actions that we perform so that we can return a response immediately to the client.
+
+These consist of:
+
+-   Validating API keys
+-   Anonymizing IPs according to project settings
+-   Decompressing and normalizing the shape of event data for the rest of the system
+-   Sending processed data to `events_plugin_ingestion` Kafka topic
+-   If any of these operations fail (other than checking for a valid API key), logging events to the Kafka `dead_letter_queue` table
+
+The goal of this step is to be as simple as possible, so that we can reliably get events into the ingestion pipeline, where Kafka can persist them until they are able to be processed.
+
+## Ingestion
+
+Due to the design of Clickhouse, update operations are slow and undesirable, and so the first 4 steps of the ingestion pipeline are performing operations on our raw events to get it ready to be persisted in Clickhouse.
+
+The ingestion pipeline is made up of a number of different step:
+
+1. [Event buffer](#1-event-buffer)
+2. [Apps - `processEvent`](#2-apps---processevent)
+3. [Person processing](#3-person-processing)
+4. [Event processing](#4-event-processing)
+5. [Writing to Clickhouse](#5-writing-to-clickhouse)
+6. [Apps - `onEvent`](#6-apps---onevent)
```

Comment:
I think we should explain more generally what the pipeline does rather than focus on order so much.

The problem is that we'll likely change it in the near future and the docs are likely to get out of date. My original plugin server docs are already outdated :(

Reply:
Updated this section to include a very high-level overview of what ingestion is. Agree, keeping things up to date is an ongoing struggle